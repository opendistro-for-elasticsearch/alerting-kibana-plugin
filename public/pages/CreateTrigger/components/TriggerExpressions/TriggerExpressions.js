/*
 *   Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *   Licensed under the Apache License, Version 2.0 (the "License").
 *   You may not use this file except in compliance with the License.
 *   A copy of the License is located at
 *
 *       http://www.apache.org/licenses/LICENSE-2.0
 *
 *   or in the "license" file accompanying this file. This file is distributed
 *   on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 *   express or implied. See the License for the specific language governing
 *   permissions and limitations under the License.
 */

import React, { Component } from 'react';
import { Field } from 'formik';
import {
  EuiExpression,
  EuiFieldNumber,
  EuiFlexGroup,
  EuiFlexItem,
  EuiFormRow,
  EuiPopover,
  EuiSelect,
} from '@elastic/eui';

const DEFAULT_CLOSED_STATES = { THRESHOLD: false };
export const Expressions = { THRESHOLD: 'THRESHOLD' };
const POPOVER_STYLE = { zIndex: '200' };
const THRESHOLD_ENUM_OPTIONS = [
  { value: 'ABOVE', text: 'IS ABOVE' },
  { value: 'BELOW', text: 'IS BELOW' },
  { value: 'EXACTLY', text: 'IS EXACTLY' },
];

class TriggerExpressions extends Component {
  constructor(props) {
    super(props);

    this.state = {
      openedStates: { ...DEFAULT_CLOSED_STATES },
    };

    this.openExpression = this.openExpression.bind(this);
    this.closeExpression = this.closeExpression.bind(this);
  }

  openExpression(expression) {
    this.setState({ openedStates: { ...DEFAULT_CLOSED_STATES, [expression]: true } });
  }

  closeExpression(expression) {
    const { openedStates } = this.state;
    this.setState({ openedStates: { ...openedStates, [expression]: false } });
  }

  renderPopover() {
    const { keyFieldName, valueFieldName } = this.props;
    return (
      <div style={POPOVER_STYLE}>
        <EuiFlexGroup
          style={{
            maxWidth: 600,
            padding: '20px',
            whiteSpace: 'nowrap',
          }}
        >
          <EuiFlexItem grow={false} style={{ width: 150 }}>
            <Field
              name={keyFieldName}
              render={({ field: { onBlur, ...rest }, form: { touched, errors } }) => (
                <EuiFormRow
                  isInvalid={touched.thresholdEnum && !!errors.thresholdEnum}
                  error={errors.thresholdEnum}
                >
                  <EuiSelect options={THRESHOLD_ENUM_OPTIONS} {...rest} />
                </EuiFormRow>
              )}
            />
          </EuiFlexItem>

          <EuiFlexItem grow={false} style={{ width: 100 }}>
            <Field
              name={valueFieldName}
              render={({ field, form: { touched, errors } }) => (
                <EuiFormRow
                  style={{ paddingLeft: '10px' }}
                  isInvalid={touched.thresholdValue && !!errors.thresholdValue}
                  error={errors.thresholdValue}
                >
                  <EuiFieldNumber {...field} />
                </EuiFormRow>
              )}
            />
          </EuiFlexItem>
        </EuiFlexGroup>
      </div>
    );
  }

  render() {
    const { openedStates } = this.state;
    const { thresholdEnum, thresholdValue, label } = this.props;
    return (
      <EuiFlexGroup alignItems="center">
        <EuiFlexItem grow={false}>
          <EuiPopover
            id="trigger-popover"
            button={
              <EuiFormRow label={label}>
                <EuiExpression
                  description={`IS ${thresholdEnum}`}
                  value={`${thresholdValue.toLocaleString()}`}
                  isActive={openedStates.THRESHOLD}
                  onClick={() => this.openExpression(Expressions.THRESHOLD)}
                />
              </EuiFormRow>
            }
            isOpen={openedStates.THRESHOLD}
            closePopover={() => this.closeExpression(Expressions.THRESHOLD)}
            panelPaddingSize="none"
            ownFocus
            withTitle
            anchorPosition="downLeft"
          >
            {this.renderPopover()}
          </EuiPopover>
        </EuiFlexItem>
      </EuiFlexGroup>
    );
  }
}

export default TriggerExpressions;
