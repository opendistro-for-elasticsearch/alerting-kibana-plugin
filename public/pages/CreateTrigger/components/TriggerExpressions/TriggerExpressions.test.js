/*
 *   Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *   Licensed under the Apache License, Version 2.0 (the "License").
 *   You may not use this file except in compliance with the License.
 *   A copy of the License is located at
 *
 *       http://www.apache.org/licenses/LICENSE-2.0
 *
 *   or in the "license" file accompanying this file. This file is distributed
 *   on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 *   express or implied. See the License for the specific language governing
 *   permissions and limitations under the License.
 */

import React from 'react';
import { mount, shallow } from 'enzyme';
import { EuiExpression } from '@elastic/eui';
import { Formik } from 'formik';

import TriggerExpressions, { Expressions } from './TriggerExpressions';

const props = {
  thresholdEnum: 'ABOVE',
  thresholdValue: 500,
};

describe('TriggerExpressions', () => {
  test('renders', () => {
    const wrapper = shallow(<TriggerExpressions {...props} />);
    expect(wrapper).toMatchSnapshot();
  });

  test('calls openExpression when clicking expression', () => {
    const wrapper = mount(<Formik render={() => <TriggerExpressions {...props} />} />);
    const openExpression = jest.spyOn(wrapper.find(TriggerExpressions).instance(), 'openExpression');
    const button = wrapper.find(EuiExpression);
    expect(wrapper.find(TriggerExpressions).state().openedStates[Expressions.THRESHOLD]).toBe(false);
    button.simulate('click');
    wrapper.update();
    expect(openExpression).toHaveBeenCalled();
    expect(wrapper.find(TriggerExpressions).state().openedStates[Expressions.THRESHOLD]).toBe(true);
  });

  test('calls closeExpression when closing popover', () => {
    const wrapper = mount(<Formik render={() => <TriggerExpressions {...props} />} />);
    const openExpression = jest.spyOn(wrapper.find(TriggerExpressions).instance(), 'openExpression');
    const closeExpression = jest.spyOn(wrapper.find(TriggerExpressions).instance(), 'closeExpression');
    const button = wrapper.find(EuiExpression);
    button.simulate('click');
    wrapper.update();
    expect(openExpression).toHaveBeenCalled();
    expect(wrapper.find(TriggerExpressions).state().openedStates[Expressions.THRESHOLD]).toBe(true);
    button.simulate('keyDown', { keyCode: 27 });
    wrapper.update();
    expect(closeExpression).toHaveBeenCalled();
    expect(wrapper.find(TriggerExpressions).state().openedStates[Expressions.THRESHOLD]).toBe(false);
  });
});