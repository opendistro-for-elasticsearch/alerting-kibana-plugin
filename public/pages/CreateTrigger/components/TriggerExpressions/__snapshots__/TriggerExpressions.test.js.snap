// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TriggerExpressions renders 1`] = `
<EuiFlexGroup
  alignItems="center"
>
  <EuiFlexItem
    grow={false}
  >
    <EuiPopover
      anchorPosition="downLeft"
      button={
        <EuiFormRow
          describedByIds={Array []}
          fullWidth={false}
          hasEmptyLabelSpace={false}
          label="Trigger condition"
          labelType="label"
        >
          <EuiExpression
            description="IS ABOVE"
            isActive={false}
            onClick={[Function]}
            value="500"
          />
        </EuiFormRow>
      }
      closePopover={[Function]}
      hasArrow={true}
      id="trigger-popover"
      isOpen={false}
      ownFocus={true}
      panelPaddingSize="none"
      withTitle={true}
    >
      <div
        style={
          Object {
            "zIndex": "200",
          }
        }
      >
        <EuiFlexGroup
          style={
            Object {
              "maxWidth": 600,
              "padding": "20px",
              "whiteSpace": "nowrap",
            }
          }
        >
          <EuiFlexItem
            grow={false}
            style={
              Object {
                "width": 150,
              }
            }
          >
            <C
              name="thresholdEnum"
              render={[Function]}
            />
          </EuiFlexItem>
          <EuiFlexItem
            grow={false}
            style={
              Object {
                "width": 100,
              }
            }
          >
            <C
              name="thresholdValue"
              render={[Function]}
            />
          </EuiFlexItem>
        </EuiFlexGroup>
      </div>
    </EuiPopover>
  </EuiFlexItem>
</EuiFlexGroup>
`;
